### PR TITLE
Bugfix: RPC/Wallet: Include HD key metadata in dumpwallet

### PR DESCRIPTION
**Pull request by luke-jr**
When restoring a backup, we should restore the wallet metadata. This moves some of it out of "comment" sections so that it can be sensibly handled by parsing tools.

(`importwallet` doesn't currently support it, but that's another matter to address)

### CODE REVIEW

File: src/wallet/rpcdump.cpp [hunk 1]
```diff
@@ -660,19 +660,26 @@ UniValue dumpwallet(const JSONRPCRequest& request)
         std::string strAddr = CBitcoinAddress(keyid).ToString();
         CKey key;
         if (pwallet->GetKey(keyid, key)) {
+            CKeyMetadata meta;
```
Comment (promag) [inline src/wallet/rpcdump.cpp]: In this case how about:
```cpp
CKeyMetadata meta = pwallet->mapKeyMetadata[keyid];
```

File: src/wallet/rpcdump.cpp [hunk 2]
```diff
@@ -660,19 +660,26 @@ UniValue dumpwallet(const JSONRPCRequest& request)
...
             } else if (mapKeyPool.count(keyid)) {
                 file << "reserve=1";
-            } else if (pwallet->mapKeyMetadata[keyid].hdKeypath == "m") {
+            } else if (meta.hdKeypath == "m") {
                 file << "inactivehdmaster=1";
             } else {
                 file << "change=1";
             }
-            file << strprintf(" # addr=%s%s\n", strAddr, (pwallet->mapKeyMetadata[keyid].hdKeypath.size() > 0 ? " hdkeypath="+pwallet->mapKeyMetadata[keyid].hdKeypath : ""));
+            file << strprintf(" # addr=%s%s\n", strAddr, (meta.hdKeypath.empty() ? "" : (" hdkeypath=" + meta.hdKeypath)));
```
Comment (promag) [inline src/wallet/rpcdump.cpp]: Nit, partial unrelated change to what commit stands for.

**Review by promag — COMMENTED** [review]
Update test? Squash 2nd and 3rd commits?

File: src/wallet/rpcdump.cpp (hunk 1 above)
Comment (luke-jr) [reply src/wallet/rpcdump.cpp]: `operator[]` can also create map items. Even if this never occurs, it needs to generate code for it.

File: src/wallet/rpcdump.cpp [hunk 3]
```diff
@@ -660,19 +660,32 @@ UniValue dumpwallet(const JSONRPCRequest& request)
...
                 file << "reserve=1";
-            } else if (pwallet->mapKeyMetadata[keyid].hdKeypath == "m") {
+            } else if (meta.hdKeypath == "m") {
                 file << "inactivehdmaster=1";
             } else {
                 file << "change=1";
             }
-            file << strprintf(" # addr=%s%s\n", strAddr, (pwallet->mapKeyMetadata[keyid].hdKeypath.size() > 0 ? " hdkeypath="+pwallet->mapKeyMetadata[keyid].hdKeypath : ""));
+            if (!meta.hdKeypath.empty()) {
+                file << " hdkeypath=" + meta.hdKeypath;
```
Comment (jonasschnelli) [inline src/wallet/rpcdump.cpp]: What is the reason to remove this out of the comment section?

**Comment by jonasschnelli** [comment]
@luke-jr: can you please elaborate a little bit what situation this PR does improve?

File: src/wallet/rpcdump.cpp (hunk 3 above)
Comment (luke-jr) [reply src/wallet/rpcdump.cpp]: It's metadata that should be restored. (our `importwallet` doesn't support it, but that's another matter)

**Review by miccomservices — APPROVED** [review]

**Comment by laanwj** [comment]
@luke-jr Can you please answer @jonasschnelli 's question? There is no rationale in the OP at all.

File: src/wallet/rpcdump.cpp (hunk 1 above)
Comment (promag) [reply src/wallet/rpcdump.cpp]: Right, off topic, but if above was changed too `const CWallet * const pwallet` then that would be false.

**Comment by luke-jr** [comment]
https://github.com/bitcoin/bitcoin/pull/11803#discussion_r154275790

**Comment by MarcoFalke** [comment]
That should probably be moved to the OP, so it is recoverable from the merge commit body. (Note that we put the OP text in the merge commit body)

File: src/wallet/rpcdump.cpp (hunk 1 above)
Comment (luke-jr) [reply src/wallet/rpcdump.cpp]: `operator[]` isn't valid at all on const maps AFAIK...?

Comment (promag) [reply src/wallet/rpcdump.cpp]: Indeed! Sorry for the noise.

**Comment by MarcoFalke** [comment]
Needs rebase if still relevant

**Comment by luke-jr** [comment]
Rebased

**Comment by MarcoFalke** [comment]
Needs rebase due to mege of #12924

File: src/wallet/rpcdump.cpp (hunk 1 above)
Comment (luke-jr) [reply src/wallet/rpcdump.cpp]: n.b. C++11 added `std::map::at`, but this isn't useful since we need to correctly handle the case where there is no metadata.

**Comment by MeshCollider** [comment]
Tested ACK https://github.com/bitcoin/bitcoin/pull/11803/commits/17d609ce26a36e0b52917f277e2b50135e405649

This isn't really a bugfix because there is no bug in the codebase which is fixed right? But I agree its a sensible change. 

Supporting it with `importwallet` is definitely something that needs to be done, ideally that should be done at the same time because they're mirror functions. But this doesn't break compatibility with the current `importwallet` so it should be ok to leave it for a future PR.

**Comment by luke-jr** [comment]
I think it's a bugfix because without it (and another fix to importwallet), dump/import will produce a wallet missing the key metadata. Isn't the goal of these RPCs to produce the same wallet?

**Comment by Sjors** [comment]
Concept ACK and lightly tested 17d609c.

Unrelated, I don't like that the address column depends on the currently active address type. We could replace it with an inferred `combo()` descriptor after #14477.

**Comment by Sjors** [comment]
This has been fixed as part of #14021 (not just for `importmulti`):
schermafbeelding 2019-02-15 om 16 03 23

It could still be useful to switch over to inferred descriptors, so we also get the origin info.

**Comment by Sjors** [comment]
Just noticed that it's a seperate field now, but sill in the comment section, so we could move it to the left.

**Comment by luke-jr** [comment]
Rebased

**Comment by Sjors** [comment]
Calling resident key metadata expert @achow101...

File: src/wallet/rpcdump.cpp [hunk 4]
```diff
@@ -839,19 +839,32 @@ UniValue dumpwallet(const JSONRPCRequest& request)
...
                 file << "reserve=1";
-            } else if (pwallet->mapKeyMetadata[keyid].hdKeypath == "s") {
+            } else if (meta.hdKeypath == "s") {
                 file << "inactivehdseed=1";
             } else {
                 file << "change=1";
             }
-            file << strprintf(" # addr=%s%s\n", strAddr, (pwallet->mapKeyMetadata[keyid].has_key_origin ? " hdkeypath="+WriteHDKeypath(pwallet->mapKeyMetadata[keyid].key_origin.path) : ""));
+            if (meta.has_key_origin) {
+                file << " hdkeypath=" + WriteHDKeypath(meta.key_origin.path);
```
Comment (achow101) [inline src/wallet/rpcdump.cpp]: We should also write out the master fingerprint here too (part of `key_origin`). I think instead of a separate field for that, we could instead write that in the same way descriptors and psbt do key origin information, i.e. fingerprint instead of `m` in the path.

**Review by achow101 — COMMENTED** [review]
Concept ACK

**Comment by prayank23** [comment]
Concept ACK

Will try this today.  Wallet with metadata is obviously better than wallet missing the key metadata. 

_This PR will complete 1337 days next week_

**Comment by prayank23** [comment]
Compiled and tried in POP!_OS with a new wallet (legacy / testnet). Diff for the dump files: https://ghostbin.com/paste/L2SXs. Observed the below things when comparing dump files for Master and PR branch:


1. `hdseedid` added for `reserve=1` key type

```diff
-cPUphC3iKWVx5HVcYimYKjvL1xn6uSHS2HKZdoLoKM1RzQkihNiD 2021-07-22T14:38:59Z reserve=1 # addr=tb1qp7m7nzhp9c4z7069m9rvrpp6tk6g4hqpj07t9l hdkeypath=m/0'/0'/8'

+cPUphC3iKWVx5HVcYimYKjvL1xn6uSHS2HKZdoLoKM1RzQkihNiD 2021-07-22T14:38:59Z reserve=1 hdkeypath=m/0'/0'/8' hdseedid=b01cf0b04e2799dab0a3658d7b59791a02c5ec86 # addr=tb1qp7m7nzhp9c4z7069m9rvrpp6tk6g4hqpj07t9l

```

2. No change in `hd master key`

```
cUEM9SuAS9z5z2bdYJhiwsmtCsq1UAwkP4nEv8i3BGK2SANwP51N 2021-07-22T14:38:59Z hdseed=1 # addr=tb1qsmkv2qs609vhhrt95wcd4xf8f6c0q89sag6fm2
```
3. No change in `script=1` key type
```
00143802740190cd3720db8d1f06f4a32f6bafc7ffad 0 script=1 # addr=2MsFRrsqi61GSNEiQKwKUFGNmzbzAevUqE7
```

**Comment by NelsonGaldeman** [comment]
I think commits should be squished.
Totally agree with prayank23: `Wallet with metadata is obviously better than wallet missing the key metadata`

This tests failed on my env though. Is it just me?
```feature_backwards_compatibility.py --descriptors   | ✖ Failed  | 1 s
feature_backwards_compatibility.py --legacy-wallet | ✖ Failed  | 1 s
feature_taproot.py --previous_release              | ✖ Failed  | 1 s
mempool_compatibility.py                           | ✖ Failed  | 1 s
rpc_createmultisig.py --legacy-wallet              | ✖ Failed  | 103 s
rpc_signrawtransaction.py --descriptors            | ✖ Failed  | 105 s
wallet_upgradewallet.py --legacy-wallet            | ✖ Failed  | 1 s
```

This is exactly how I fetched the branch, compiled the binnaries and ran the tests.
```
git fetch core
git fetch core pull/11803/merge:pr11803
git switch pr11803
./autogen.sh
./configure -q --without-gui
make clean > /dev/null
make -j 4
python3 test/functional/test_runner.py
 ```
 
Checked my HEAD commit within my local PR copy and I'm getting 8642fbd04ce670cb76657580990d4073e41b0fe1 which differs from the last commit I see on the commit tab here, am I doing something wrong?

**Comment by Rspigler** [comment]
Tested on master for legacy wallets:

I get an extended private masterkey

Then a list of (private key, address, hdkeypath); although very out of order.  Somewhere in there, is the hdseed

Master for descriptor wallets:

> This type of wallet does not support this command (code -4)

Tested on 563dda4b52aca970189b757b7a86615adc442608:

Legacy:

Extended private master key.

Then list of (private key, hdkeypath, hdseedid, address).  Changes are address is moved to the end of the list, and addition of hdseedid.   The keys are still out of order, and the hdseed is also still randomly in the list.

Descriptor:

Unable to create descriptor wallet (either through the GUI or console), so unable to test.


Possible improvements?:

- [ ] Work with descriptors
- [ ] Sort addresses/hdkeypaths by address index
- [ ] Bring the hdseed to top of list, like the extended private master key is
- [ ] List the hdseedid once at top of list, instead of repeated for every address?

**Comment by Rspigler** [comment]
>  Work with descriptors

Maybe I was thinking of `listdescriptors`
